dipy/denoise/localpca.py

```diff
@@ -0,0 +1,171 @@
+import numpy as np
+import scipy as sp
+
+
+def localpca(arr, sigma, patch_radius=1, tou=0, rician=True):
+    '''
+    Local PCA Based Denoising of Diffusion Datasets
+
+    References
+    ----------
+    Diffusion Weighted Image Denoising Using Overcomplete Local PCA
+    Manjon JV, Coupe P, Concha L, Buades A, Collins DL
+
+    Parameters
+    ----------
+    arr : A 4D array which is to be denoised
+    sigma : float or 3D array
+        standard deviation of the noise estimated from the data
+    patch_radius : The radius of the local patch to
+                be taken around each voxel
+    tou : float or 3D array
+        threshold parameter for diagonal matrix thresholding,
+        default value = (2.3 * sigma * sigma)
+    rician : boolean
+        If True the noise is estimated as Rician, otherwise Gaussian noise
+        is assumed.
+
+    Returns
+    -------
+
+    denoised_arr : 4D array
+        this is the denoised array of the same size as that of
+        arr (input data)
+
+    '''
+
+    # Read the array and fix the dimensions
+
+    if arr.ndim == 4:
+        sigma = np.zeros(arr[..., 0].shape, dtype=np.float64)
+
+        if tou == 0:
+            tou = 2.3 * 2.3 * sigma
+
+        if isinstance(sigma, np.ndarray) and sigma.ndim == 3:
+
+            sigma = (np.ones(arr.shape, dtype=np.float64)
+                     * sigma[..., np.newaxis])
+            tou = (np.ones(arr.shape, dtype=np.float64) * tou[..., np.newaxis])
+        else:
+            sigma = np.ones(arr.shape, dtype=np.float64) * sigma
+            tou = np.ones(arr.shape, dtype=np.float64) * tou
+
+        # loop around and find the 3D patch for each direction at each pixel
+
+        # declare arrays for theta and thetax
+        theta = np.zeros(arr.shape, dtype=np.float64)
+        thetax = np.zeros(arr.shape, dtype=np.float64)
+
+        patch_size = 2 * patch_radius + 1
+
+        for k in range(patch_radius, arr.shape[2] - patch_radius, 1):
+            print(k)
+            for j in range(patch_radius, arr.shape[1] - patch_radius, 1):
+                for i in range(patch_radius, arr.shape[0] - patch_radius, 1):
+
+                    X = np.zeros(
+                        (patch_size * patch_size * patch_size, arr.shape[3]))
+                    M = np.zeros(arr.shape[3])
+
+                    temp = arr[
+                        i -
+                        patch_radius: i +
+                        patch_radius +
+                        1,
+                        j -
+                        patch_radius: j +
+                        patch_radius +
+                        1,
+                        k -
+                        patch_radius: k +
+                        patch_radius +
+                        1,
+                        :]
+                    X = temp.reshape(
```

Comment:
Sorry, this maybe a dumb question but I always get confused with this kind of reshapes: are we sure that after this reshape, each **row** of X has one of the diffusion signals within the local window?